kubernetes/client.go

```diff
@@ -123,34 +125,128 @@ func (in *IstioClient) GetServices(namespaceName string) (*v1.ServiceList, error
-	if err != nil {
-		return nil, err
+	serviceChan, endpointsChan := make(chan *v1.Service), make(chan *v1.Endpoints)
+	errChan := make(chan error)
+
+	go func() {
+		service, err := in.k8s.CoreV1().Services(namespace).Get(serviceName, emptyGetOptions)
+		if err != nil {
+			errChan <- err
+		} else {
+			serviceChan <- service
+		}
+	}()
+
+	go func() {
+		endpoints, err := in.k8s.CoreV1().Endpoints(namespace).Get(serviceName, emptyGetOptions)
+		if err != nil {
+			errChan <- err
+		} else {
+			endpointsChan <- endpoints
+		}
+	}()
+
+	var serviceDetails = ServiceDetails{}
+	var serviceDetailsError error
+
+	select {
+	case serviceDetails.Service = <-serviceChan:
+	case serviceDetailsError = <-errChan:
 	}
-	endpoints, err := in.k8s.CoreV1().Endpoints(namespace).Get(serviceName, emptyGetOptions)
-	if err != nil {
-		return nil, err
+	select {
+	case serviceDetails.Endpoints = <-endpointsChan:
+	case serviceDetailsError = <-errChan:
+	}
+
+	if serviceDetailsError != nil {
+		return nil, serviceDetailsError
 	}
-	pods := make([]*v1.Pod, 0)
-	for _, subset := range endpoints.Subsets {
+
+	podNames := make([]string, 0)
+	for _, subset := range serviceDetails.Endpoints.Subsets {
 		for _, address := range subset.Addresses {
 			targetRef := address.TargetRef
 			if targetRef != nil && targetRef.Kind == "Pod" {
-				pod, err := in.k8s.CoreV1().Pods(namespace).Get(targetRef.Name, emptyGetOptions)
-				if err != nil {
-					return nil, err
-				}
-				pods = append(pods, pod)
+				podNames = append(podNames, targetRef.Name)
 			}
 		}
 	}
-	return &ServiceDetails{service, endpoints, pods}, nil
+
+	podChan, podErrChan := make(chan *v1.Pod, len(podNames)), make(chan error, len(podNames))
+	var wg sync.WaitGroup
+
+	for _, podName := range podNames {
+		wg.Add(1)
+		go func(podName string) {
+			defer wg.Done()
+			log.Infof("podName %v", podName)
+			pod, err := in.k8s.CoreV1().Pods(namespace).Get(podName, emptyGetOptions)
+			if err != nil {
+				podErrChan <- err
+			} else {
+				podChan <- pod
+			}
+		}(podName)
+	}
+
+	go func() {
+		wg.Wait()
+		close(podChan)
+		close(podErrChan)
+	}()
+
+	for podErr := range podErrChan {
+		return nil, podErr
+	}
+
+	serviceDetails.Pods = make([]*v1.Pod, 0)
+	for pod := range podChan {
+		serviceDetails.Pods = append(serviceDetails.Pods, pod)
+	}
+
+	return &serviceDetails, nil
 }
 
-// GetIstioDetails returns Istio details for a given service, on this version it describes the RouterRules defined for a service.
+// GetIstioDetails returns Istio details for a given service,
+// on this version it describes the RouterRules and DestinationPolicies defined for a service.
 // A service is defined by the namespace and the service name.
 // It returns an error on any problem.
 func (in *IstioClient) GetIstioDetails(namespace string, serviceName string) (*IstioDetails, error) {
+	routerRulesChan, destinationPoliciesChan := make(chan []*RouteRule), make(chan []*DestinationPolicy)
+	errChan := make(chan error)
```

Comment:
Just as a comment, not even a suggestion : I've found another pattern to deal with errors with channels, while reading sites and blogs... It's simply to embed the error in the model. It reduces boilerplate code quite a lot (you can see an example here: https://github.com/swift-sunshine/swscore/pull/94/files#diff-ad09977461f7b9ea7927917031ae47cbR81 )
Then error in model may or may not be exported, depending on where you want to handle it, either somewhere in the same package or forwarded up to the frontend.

Reply:
I see, so the idea is to use a private/temporal type to group return and error and just use a single channel, as response is always there.
I don't have a strong opinion, so I can follow that for this one.

Reply:
Yes, you can try and see what you prefer, in my case I found it more readable & simpler, but perhaps it won't apply for every case.